staging/src/k8s.io/legacy-cloud-providers/azure/azure_vmss.go

```diff
@@ -780,6 +813,137 @@ func (ss *scaleSet) EnsureHostInPool(service *v1.Service, nodeName types.NodeNam
+				continue
+			}
+			// in this scenario the vmSetName is an empty string and the name of vmss should be obtained from the provider IDs of nodes
+			vmssName, resourceGroupName, err := getVmssAndResourceGroupNameByVMProviderID(node.Spec.ProviderID)
+			if err != nil {
+				return err
+			}
+			// only vmsses in the resource group same as it's in azure config are included
+			if resourceGroupName == ss.ResourceGroup {
+				vmssNamesMap[vmssName] = true
+			}
+		}
+	} else {
+		vmssNamesMap[vmSetName] = true
+	}
+
+	for vmssName := range vmssNamesMap {
+		vmss, err := ss.GetScaleSetWithRetry(service, ss.ResourceGroup, vmssName)
+		if err != nil {
+			return err
+		}
+		vmssNIC := vmss.VirtualMachineProfile.NetworkProfile.NetworkInterfaceConfigurations
+		primaryNIC, err := ss.getPrimaryNetworkInterfaceConfigurationForScaleSet(*vmssNIC, vmssName)
+		if err != nil {
+			return err
+		}
+		primaryIPConfig, err := ss.getPrimaryIPConfigForScaleSet(primaryNIC, vmssName)
+		if err != nil {
+			return err
+		}
+		loadBalancerBackendAddressPools := []compute.SubResource{}
+		if primaryIPConfig.LoadBalancerBackendAddressPools != nil {
+			loadBalancerBackendAddressPools = *primaryIPConfig.LoadBalancerBackendAddressPools
+		}
+
+		var found bool
+		for _, loadBalancerBackendAddressPool := range loadBalancerBackendAddressPools {
+			if strings.EqualFold(*loadBalancerBackendAddressPool.ID, backendPoolID) {
+				found = true
+				break
+			}
+		}
+		if found {
+			continue
+		}
+
+		if ss.useStandardLoadBalancer() && len(loadBalancerBackendAddressPools) > 0 {
+			// Although standard load balancer supports backends from multiple scale
+			// sets, the same network interface couldn't be added to more than one load balancer of
+			// the same type. Omit those nodes (e.g. masters) so Azure ARM won't complain
+			// about this.
+			newBackendPoolsIDs := make([]string, 0, len(loadBalancerBackendAddressPools))
+			for _, pool := range loadBalancerBackendAddressPools {
+				if pool.ID != nil {
+					newBackendPoolsIDs = append(newBackendPoolsIDs, *pool.ID)
+				}
+			}
+			isSameLB, oldLBName, err := isBackendPoolOnSameLB(backendPoolID, newBackendPoolsIDs)
+			if err != nil {
+				return err
+			}
+			if !isSameLB {
+				klog.V(4).Infof("VMSS %q has already been added to LB %q, omit adding it to a new one", vmssName, oldLBName)
+				return nil
+			}
+		}
+
+		// Compose a new vmss with added backendPoolID.
+		loadBalancerBackendAddressPools = append(loadBalancerBackendAddressPools,
+			compute.SubResource{
+				ID: to.StringPtr(backendPoolID),
+			})
+		newVMSSUpdate := compute.VirtualMachineScaleSetUpdate{
+			VirtualMachineScaleSetUpdateProperties: &compute.VirtualMachineScaleSetUpdateProperties{
+				VirtualMachineProfile: &compute.VirtualMachineScaleSetUpdateVMProfile{
+					NetworkProfile: &compute.VirtualMachineScaleSetUpdateNetworkProfile{
+						NetworkInterfaceConfigurations: &[]compute.VirtualMachineScaleSetUpdateNetworkConfiguration{
```

Comment:
If there're multiple network interface configurations in the VMSS, this change would delete all others and only keep the primary one. So we need to a list of original list, change the primary one and put again here.

Reply:
Actually, the struct types are different: `VirtualMachineScaleSetUpdateNetworkProfile` versus `VirtualMachineScaleSetNetworkConfiguration`. Hence the simple "update and put back" doesn't work. I'm wondering if there's a good solution under this circumstance.

Reply:
I believe a deep copy is needed here. WDYT?

Reply:
Have push another commit to user `CreateOrUpdate` instead of `Update`